File: backend/app/api/transactions.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.models.models import Transaction as TransactionModel, User as UserModel, Expense as ExpenseModel
from app.schemas.schemas import TransactionCreate, TransactionUpdate, TransactionResponse
from app.services.finance import calculate_transaction
from app.services.utils import get_current_week_id, validate_day
from app.api.deps import check_role, get_current_user
from app.models.models import UserRole, TipMethod
from sqlalchemy import func
from typing import List

router = APIRouter()

@router.get("/", response_model=None)
def get_transactions(
    day: str = None,
    week_id: str = None,
    washer_id: int = None,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user)
):
    validate_day(day)
    query = db.query(TransactionModel)
    if day:
        query = query.filter(func.date(TransactionModel.timestamp) == day)
    elif week_id:
        query = query.filter(TransactionModel.week_id == week_id)

    if washer_id:
        query = query.filter(TransactionModel.washer_id == washer_id)

    # Employees can only see their own
    if current_user.role == UserRole.EMPLOYEE:
        query = query.filter(TransactionModel.washer_id == current_user.id)

    return query.all()

def _reverse_balances(tx: TransactionModel, db: Session):
    """Reverse the balance effects of a transaction on its washer."""
    washer = db.query(UserModel).filter(UserModel.id == tx.washer_id).first()
    if washer:
        # The transaction added credit_wages to payable and debit_debt to debt
        # We stored: final_payout (negative = debt, positive = wage)
        credit = max(0.0, tx.final_payout)
        debit = abs(min(0.0, tx.final_payout))
        washer.payable_balance -= credit
        washer.debt_balance -= debit

def _apply_balances(tx: TransactionModel, db: Session):
    """Apply balance effects of a transaction on its washer."""
    washer = db.query(UserModel).filter(UserModel.id == tx.washer_id).first()
    if washer:
        credit = max(0.0, tx.final_payout)
        debit = abs(min(0.0, tx.final_payout))
        washer.payable_balance += credit
        washer.debt_balance += debit

def _sync_chai_expense(tx: TransactionModel, db: Session, isolated_tip: float, washer_name: str):
    """Keep the cash-tip 'Chai' expense in sync with a transaction.
       A tip isolated with CASH method is handed to the attendant, so the
       business must record it as an expense (float parity)."""
    existing = db.query(ExpenseModel).filter(ExpenseModel.transaction_id == tx.id).all()
    for exp in existing:
        db.delete(exp)
    if isolated_tip > 0 and tx.tip_method.value == TipMethod.CASH:
        db.add(ExpenseModel(
            description=f"Chai (tip) - {washer_name}",
            amount=isolated_tip,
            category="Chai",
            week_id=tx.week_id,
            transaction_id=tx.id
        ))

def _validate_misc(data: TransactionCreate):
    if data.misc_amount and data.misc_amount > 0:
        extra = max(0.0, (data.cash_paid + data.mpesa_paid) - data.expected_price)
        if data.misc_amount > extra:
            raise HTTPException(
                status_code=400,
                detail=f"Miscellaneous amount ({data.misc_amount}) exceeds overpayment ({extra}). "
                       "Amount beyond expected must be tip or miscellaneous."
            )
    if data.misc_amount and data.misc_amount > 0 and not data.misc_description:
        raise HTTPException(status_code=400, detail="Miscellaneous amount requires a description")

def _build_create_from_update(tx: TransactionModel, update: TransactionUpdate) -> TransactionCreate:
    """Merge existing transaction with update fields into a TransactionCreate."""
    return TransactionCreate(
        washer_id=update.washer_id if update.washer_id is not None else tx.washer_id,
        category=update.category if update.category is not None else tx.category,
        expected_price=update.expected_price if update.expected_price is not None else tx.expected_price,
        cash_paid=update.cash_paid if update.cash_paid is not None else tx.cash_paid,
        mpesa_paid=update.mpesa_paid if update.mpesa_paid is not None else tx.mpesa_paid,
        mpesa_transaction_id=update.mpesa_transaction_id if update.mpesa_transaction_id is not None else tx.mpesa_transaction_id,
        mpesa_sender_name=update.mpesa_sender_name if update.mpesa_sender_name is not None else tx.mpesa_sender_name,
        manual_tip=update.manual_tip if update.manual_tip is not None else tx.manual_tip,
        tip_method=update.tip_method if update.tip_method is not None else tx.tip_method,
        misc_amount=update.misc_amount if update.misc_amount is not None else tx.misc_amount or 0.0,
        misc_description=update.misc_description if update.misc_description is not None else tx.misc_description,
        has_car_wash=update.has_car_wash if update.has_car_wash is not None else tx.has_car_wash,
        has_vacuum=update.has_vacuum if update.has_vacuum is not None else tx.has_vacuum,
        has_engine_wash=update.has_engine_wash if update.has_engine_wash is not None else tx.has_engine_wash,
        plate_number=update.plate_number if update.plate_number is not None else tx.plate_number,
        custom_category=update.custom_category if update.custom_category is not None else tx.custom_category,
        carpet_metadata=update.carpet_metadata,
    )

@router.post("/", response_model=TransactionResponse)
def create_transaction(
    transaction_in: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))
):
    result = calculate_transaction(transaction_in)
    _validate_misc(transaction_in)
    week_id = get_current_week_id()

    db_transaction = TransactionModel(
        washer_id=transaction_in.washer_id,
        category=transaction_in.category,
        expected_price=result.transaction_summary.expected_price,
        cash_paid=transaction_in.cash_paid,
        mpesa_paid=transaction_in.mpesa_paid,
        mpesa_transaction_id=transaction_in.mpesa_transaction_id,
        mpesa_sender_name=transaction_in.mpesa_sender_name,
        manual_tip=transaction_in.manual_tip,
        tip_method=transaction_in.tip_method,
        misc_amount=transaction_in.misc_amount or 0.0,
        misc_description=transaction_in.misc_description,
        has_car_wash=transaction_in.has_car_wash,
        has_vacuum=transaction_in.has_vacuum,
        has_engine_wash=transaction_in.has_engine_wash,
        plate_number=transaction_in.plate_number,
        custom_category=transaction_in.custom_category,
        customer_phone=transaction_in.carpet_metadata.customer_phone if transaction_in.carpet_metadata else None,
        carpet_characteristics=transaction_in.carpet_metadata.characteristics if transaction_in.carpet_metadata else None,
        receiver_id=transaction_in.carpet_metadata.receiver_id if transaction_in.carpet_metadata else None,
        total_paid=result.transaction_summary.total_customer_paid,
        isolated_tip=result.transaction_summary.isolated_tip,
        net_business_remittance=result.transaction_summary.net_business_remittance,
        shortfall=result.transaction_summary.shortfall_detected,
        calculated_commission=result.employee_financials.calculated_commission,
        net_wage_before_tip=result.employee_financials.net_wage_before_tip,
        final_payout=result.employee_financials.final_payout_output,
        week_id=week_id
    )

    db.add(db_transaction)

    washer = db.query(UserModel).filter(UserModel.id == transaction_in.washer_id).first()
    if not washer:
        raise HTTPException(status_code=404, detail="Washer not found")

    washer.payable_balance += result.ledger_routing.credit_employee_wages
    washer.debt_balance += result.ledger_routing.debit_employee_debt

    db.commit()
    db.refresh(db_transaction)

    _sync_chai_expense(db_transaction, db, result.transaction_summary.isolated_tip, washer.full_name)
    db.commit()

    return TransactionResponse(
        id=db_transaction.id,
        timestamp=db_transaction.timestamp,
        transaction_summary=result.transaction_summary,
        employee_financials=result.employee_financials,
        ledger_routing=result.ledger_routing
    )

@router.put("/{transaction_id}", response_model=TransactionResponse)
def update_transaction(
    transaction_id: int,
    update: TransactionUpdate,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))
):
    tx = db.query(TransactionModel).filter(TransactionModel.id == transaction_id).first()
    if not tx:
        raise HTTPException(status_code=404, detail="Transaction not found")

    # Reverse old balances
    _reverse_balances(tx, db)

    # Build merged create data and recalculate
    create_data = _build_create_from_update(tx, update)
    _validate_misc(create_data)
    result = calculate_transaction(create_data)

    # Check washer exists
    washer = db.query(UserModel).filter(UserModel.id == create_data.washer_id).first()
    if not washer:
        raise HTTPException(status_code=404, detail="Washer not found")

    # Update transaction fields
    tx.washer_id = create_data.washer_id
    tx.category = create_data.category
    tx.expected_price = result.transaction_summary.expected_price
    tx.cash_paid = create_data.cash_paid
    tx.mpesa_paid = create_data.mpesa_paid
    tx.mpesa_transaction_id = create_data.mpesa_transaction_id
    tx.mpesa_sender_name = create_data.mpesa_sender_name
    tx.manual_tip = create_data.manual_tip
    tx.tip_method = create_data.tip_method
    tx.misc_amount = create_data.misc_amount or 0.0
    tx.misc_description = create_data.misc_description
    tx.has_car_wash = create_data.has_car_wash
    tx.has_vacuum = create_data.has_vacuum
    tx.has_engine_wash = create_data.has_engine_wash
    tx.plate_number = create_data.plate_number
    tx.custom_category = create_data.custom_category
    tx.customer_phone = create_data.carpet_metadata.customer_phone if create_data.carpet_metadata else None
    tx.carpet_characteristics = create_data.carpet_metadata.characteristics if create_data.carpet_metadata else None
    tx.receiver_id = create_data.carpet_metadata.receiver_id if create_data.carpet_metadata else None
    tx.total_paid = result.transaction_summary.total_customer_paid
    tx.isolated_tip = result.transaction_summary.isolated_tip
    tx.net_business_remittance = result.transaction_summary.net_business_remittance
    tx.shortfall = result.transaction_summary.shortfall_detected
    tx.calculated_commission = result.employee_financials.calculated_commission
    tx.net_wage_before_tip = result.employee_financials.net_wage_before_tip
    tx.final_payout = result.employee_financials.final_payout_output

    # Apply new balances
    _apply_balances(tx, db)

    _sync_chai_expense(tx, db, result.transaction_summary.isolated_tip, washer.full_name)

    db.commit()
    db.refresh(tx)

    return TransactionResponse(
        id=tx.id,
        timestamp=tx.timestamp,
        transaction_summary=result.transaction_summary,
        employee_financials=result.employee_financials,
        ledger_routing=result.ledger_routing
    )

@router.delete("/{transaction_id}")
def delete_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(check_role([UserRole.ADMIN, UserRole.MANAGER]))
):
    tx = db.query(TransactionModel).filter(TransactionModel.id == transaction_id).first()
    if not tx:
        raise HTTPException(status_code=404, detail="Transaction not found")

    _reverse_balances(tx, db)
    for exp in db.query(ExpenseModel).filter(ExpenseModel.transaction_id == tx.id).all():
        db.delete(exp)
    db.delete(tx)
    db.commit()

    return {"message": "Transaction deleted"}

File: server/src/routes/transactions.ts
import { Router } from 'express';
import { pool, queryOne } from '../db';
import { HttpError, getCurrentWeekId, validateDay, wrap } from '../utils';
import { asFloat, asInt, asBool, asOptFloat, asOptInt, asOptStr, requireEnum, calculateTransaction } from '../finance';
import { TransactionSummary, EmployeeFinancials, LedgerRouting } from '../finance';
import { getCurrentUser, requireRole, AuthRequest } from '../auth';
import { SERVICE_CATEGORIES, TIP_METHODS } from '../types';
import type { UserRow, TxRow, ExpenseRow } from '../types';
import { iso, serializeTx } from '../serialize';
import { ROLES } from '../types';

const router = Router();

export interface CarpetMetadataIn {
  characteristics: string;
  receiver_id: number;
  customer_phone: string | null;
}

export interface CreateData {
  washer_id: number;
  category: string;
  expected_price: number;
  cash_paid: number;
  mpesa_paid: number;
  mpesa_transaction_id: string | null;
  mpesa_sender_name: string | null;
  manual_tip: number;
  tip_method: string;
  misc_amount: number;
  misc_description: string | null;
  has_car_wash: boolean;
  has_vacuum: boolean;
  has_engine_wash: boolean;
  plate_number: string | null;
  custom_category: string | null;
  carpet_metadata: CarpetMetadataIn | null;
}

function parseMetadata(v: unknown): CarpetMetadataIn | null {
  if (v === undefined || v === null) return null;
  if (typeof v !== 'object' || Array.isArray(v)) throw new HttpError(422, 'carpet_metadata must be an object');
  const o = v as Record<string, unknown>;
  if (o.characteristics === undefined || o.receiver_id === undefined) {
    throw new HttpError(422, 'carpet_metadata requires characteristics and receiver_id');
  }
  return {
    characteristics: asOptStr(o.characteristics) ?? '',
    receiver_id: asInt(o.receiver_id, 'receiver_id'),
    customer_phone: asOptStr(o.customer_phone),
  };
}

export function parseCreate(body: Record<string, unknown>): CreateData {
  const data: CreateData = {
    washer_id: asInt(body.washer_id, 'washer_id'),
    category: requireEnum(asOptStr(body.category) ?? '', SERVICE_CATEGORIES, 'category'),
    expected_price: asFloat(body.expected_price, 'expected_price'),
    cash_paid: asFloat(body.cash_paid, 'cash_paid'),
    mpesa_paid: asFloat(body.mpesa_paid, 'mpesa_paid'),
    mpesa_transaction_id: asOptStr(body.mpesa_transaction_id),
    mpesa_sender_name: asOptStr(body.mpesa_sender_name),
    manual_tip: asOptFloat(body.manual_tip, 'manual_tip') ?? 0,
    tip_method: requireEnum(asOptStr(body.tip_method) ?? '', TIP_METHODS, 'tip_method'),
    misc_amount: asOptFloat(body.misc_amount, 'misc_amount') ?? 0,
    misc_description: asOptStr(body.misc_description),
    has_car_wash: body.has_car_wash === undefined ? false : asBool(body.has_car_wash, 'has_car_wash'),
    has_vacuum: body.has_vacuum === undefined ? false : asBool(body.has_vacuum, 'has_vacuum'),
    has_engine_wash: body.has_engine_wash === undefined ? false : asBool(body.has_engine_wash, 'has_engine_wash'),
    plate_number: asOptStr(body.plate_number),
    custom_category: asOptStr(body.custom_category),
    carpet_metadata: parseMetadata(body.carpet_metadata),
  };
  for (const k of ['expected_price', 'cash_paid', 'mpesa_paid', 'manual_tip', 'misc_amount']) {
    if ((data as any)[k] < 0) throw new HttpError(422, `${k} must be >= 0`);
  }
  return data;
}

export function validateMisc(data: CreateData): void {
  const extra = Math.max(0.0, data.cash_paid + data.mpesa_paid - data.expected_price);
  const wanted = data.misc_amount;
  if (wanted && wanted > extra) {
    throw new HttpError(
      400,
      `Miscellaneous amount (${wanted}) exceeds overpayment (${extra}). Amount beyond expected must be tip or miscellaneous.`,
    );
  }
  if (wanted && !data.misc_description) {
    throw new HttpError(400, 'Miscellaneous amount requires a description');
  }
}

const TX_FIELDS =
  `id, timestamp, washer_id, category, expected_price, cash_paid, mpesa_paid, mpesa_transaction_id, ` +
  `mpesa_sender_name, manual_tip, tip_method, misc_amount, misc_description, has_car_wash, has_vacuum, ` +
  `has_engine_wash, plate_number, customer_phone, custom_category, carpet_characteristics, receiver_id, ` +
  `total_paid, isolated_tip, net_business_remittance, shortfall, calculated_commission, ` +
  `net_wage_before_tip, final_payout, week_id`;

export interface CreateResult {
  id: number;
  timestamp: string;
  transaction_summary: TransactionSummary;
  employee_financials: EmployeeFinancials;
  ledger_routing: LedgerRouting;
}

// Shared core used by POST /transactions/ and the carpet release flow.
export async function createTransactionCore(data: CreateData, actor: UserRow): Promise<CreateResult> {
  const result = calculateTransaction(data);
  validateMisc(data);

  const washer = await queryOne<UserRow>('SELECT * FROM users WHERE id = $1', [data.washer_id]);
  if (!washer) throw new HttpError(404, 'Washer not found');

  const md = data.carpet_metadata;
  const week_id = getCurrentWeekId();
  const now = new Date();

  const client = await pool.connect();
  try {
    await client.query('BEGIN');
    const insert = await client.query(
      `INSERT INTO transactions (
        timestamp, washer_id, category, expected_price, cash_paid, mpesa_paid, mpesa_transaction_id,
        mpesa_sender_name, manual_tip, tip_method, misc_amount, misc_description, has_car_wash, has_vacuum,
        has_engine_wash, plate_number, customer_phone, custom_category, carpet_characteristics, receiver_id,
        total_paid, isolated_tip, net_business_remittance, shortfall, calculated_commission,
        net_wage_before_tip, final_payout, week_id
      ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24,$25,$26,$27,$28)
      RETURNING ${TX_FIELDS}`,
      [
        now.toISOString().replace('T', ' ').replace('Z', ''),
        data.washer_id,
        data.category.toUpperCase(),
        result.transaction_summary.expected_price,
        data.cash_paid,
        data.mpesa_paid,
        data.mpesa_transaction_id,
        data.mpesa_sender_name,
        data.manual_tip,
        data.tip_method.toUpperCase(),
        data.misc_amount || 0,
        data.misc_description,
        data.has_car_wash,
        data.has_vacuum,
        data.has_engine_wash,
        data.plate_number,
        md?.customer_phone ?? null,
        data.custom_category,
        md?.characteristics ?? null,
        md?.receiver_id ?? null,
        result.transaction_summary.total_customer_paid,
        result.transaction_summary.isolated_tip,
        result.transaction_summary.net_business_remittance,
        result.transaction_summary.shortfall_detected,
        result.employee_financials.calculated_commission,
        result.employee_financials.net_wage_before_tip,
        result.employee_financials.final_payout_output,
        week_id,
      ],
    );
    const tx = insert.rows[0] as TxRow;

    const credit = result.ledger_routing.credit_employee_wages;
    const debit = result.ledger_routing.debit_employee_debt;
    await client.query(
      'UPDATE users SET payable_balance = payable_balance + $1, debt_balance = debt_balance + $2 WHERE id = $3',
      [credit, debit, data.washer_id],
    );

    await syncChaiExpense(client, tx, result.transaction_summary.isolated_tip, washer.full_name ?? '');
    await client.query('COMMIT');

    return {
      id: tx.id,
      timestamp: iso(tx.timestamp) ?? '',
      transaction_summary: result.transaction_summary,
      employee_financials: result.employee_financials,
      ledger_routing: result.ledger_routing,
    };
  } catch (e) {
    await client.query('ROLLBACK');
    throw e;
  } finally {
    client.release();
  }
}

// Mirrors _sync_chai_expense: for CASH-isolated tips, record a "Chai (tip)" expense linked to the tx.
async function syncChaiExpense(
  client: any,
  tx: TxRow,
  isolated_tip: number,
  washerName: string,
): Promise<void> {
  await client.query('DELETE FROM expenses WHERE transaction_id = $1', [tx.id]);
  if (isolated_tip > 0 && tx.tip_method === 'CASH') {
    const now = new Date().toISOString().replace('T', ' ').replace('Z', '');
    await client.query(
      `INSERT INTO expenses (timestamp, description, amount, category, week_id, transaction_id)
       VALUES ($1,$2,$3,$4,$5,$6)`,
      [now, `Chai (tip) - ${washerName}`, isolated_tip, 'Chai', tx.week_id, tx.id],
    );
  }
}

async function reverseBalances(client: any, tx: TxRow): Promise<void> {
  const credit = Math.max(0.0, tx.final_payout);
  const debit = Math.abs(Math.min(0.0, tx.final_payout));
  await client.query(
    'UPDATE users SET payable_balance = payable_balance - $1, debt_balance = debt_balance - $2 WHERE id = $3',
    [credit, debit, tx.washer_id],
  );
}

// GET /transactions
router.get(
  '/',
  wrap(async (req: AuthRequest, res) => {
    validateDay((req.query.day as string) ?? null);
    const day = req.query.day as string | undefined;
    const week_id = req.query.week_id as string | undefined;
    const washer_id = req.query.washer_id as string | undefined;
    const user = await getCurrentUser(req);

    const conds: string[] = [];
    const params: any[] = [];
    if (day) {
      params.push(day);
      conds.push(`date(timestamp) = $${params.length}::date`);
    } else if (week_id) {
      params.push(week_id);
      conds.push(`week_id = $${params.length}`);
    }
    if (washer_id) {
      const wid = Number(washer_id);
      if (!Number.isInteger(wid)) throw new HttpError(422, 'washer_id must be an integer');
      params.push(wid);
      conds.push(`washer_id = $${params.length}`);
    }
    if (user.role.toUpperCase() === 'EMPLOYEE') {
      params.push(user.id);
      conds.push(`washer_id = $${params.length}`);
    }

    const rows = await (
      await pool.query(
        `SELECT ${TX_FIELDS} FROM transactions ${conds.length ? 'WHERE ' + conds.join(' AND ') : ''} ORDER BY timestamp DESC`,
        params,
      )
    ).rows;
    res.json(rows.map(serializeTx));
  }),
);

// POST /transactions
router.post(
  '/',
  wrap(async (req: AuthRequest, res) => {
    await requireRole(req, ROLES.slice(0, 2)); // ADMIN, MANAGER
    const data = parseCreate(req.body || {});
    const result = await createTransactionCore(data, req.user!);
    res.status(200).json(result);
  }),
);

// PUT /transactions/:id
router.put(
  '/:id',
  wrap(async (req: AuthRequest, res) => {
    await requireRole(req, ROLES.slice(0, 2));
    const txId = Number(req.params.id);
    if (!Number.isInteger(txId)) throw new HttpError(404, 'Transaction not found');

    const body = req.body || {};
    const updates = (() => {
      const u: Record<string, any> = {};
      for (const k of [
        'washer_id', 'category', 'expected_price', 'cash_paid', 'mpesa_paid', 'mpesa_transaction_id',
        'mpesa_sender_name', 'manual_tip', 'tip_method', 'misc_amount', 'misc_description', 'has_car_wash',
        'has_vacuum', 'has_engine_wash', 'plate_number', 'custom_category', 'carpet_metadata',
      ]) {
        if (body[k] !== undefined) u[k] = body[k];
      }
      return u;
    })();
    const notPresent = (k: string) => updates[k] === undefined;

    const tx = await queryOne<TxRow>(`SELECT ${TX_FIELDS} FROM transactions WHERE id = $1`, [txId]);
    if (!tx) throw new HttpError(404, 'Transaction not found');

    const merged = mergeUpdate(tx, updates);
    if (updates.carpet_metadata !== undefined) {
      merged.carpet_metadata = parseMetadata(updates.carpet_metadata);
    }
    const result = calculateTransaction(merged);
    validateMisc(merged);
    const washer = await queryOne<UserRow>('SELECT * FROM users WHERE id = $1', [merged.washer_id]);
    if (!washer) throw new HttpError(404, 'Washer not found');

    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      await reverseBalances(client, tx);

      const md = merged.carpet_metadata;
      await client.query(
        `UPDATE transactions SET
          washer_id=$1, category=$2, expected_price=$3, cash_paid=$4, mpesa_paid=$5, mpesa_transaction_id=$6,
          mpesa_sender_name=$7, manual_tip=$8, tip_method=$9, misc_amount=$10, misc_description=$11,
          has_car_wash=$12, has_vacuum=$13, has_engine_wash=$14, plate_number=$15, customer_phone=$16,
          custom_category=$17, carpet_characteristics=$18, receiver_id=$19, total_paid=$20, isolated_tip=$21,
          net_business_remittance=$22, shortfall=$23, calculated_commission=$24, net_wage_before_tip=$25,
          final_payout=$26
         WHERE id=$27`,
        [
          merged.washer_id,
          merged.category.toUpperCase(),
          result.transaction_summary.expected_price,
          merged.cash_paid,
          merged.mpesa_paid,
          merged.mpesa_transaction_id,
          merged.mpesa_sender_name,
          merged.manual_tip,
          merged.tip_method.toUpperCase(),
          merged.misc_amount || 0,
          merged.misc_description,
          merged.has_car_wash,
          merged.has_vacuum,
          merged.has_engine_wash,
          merged.plate_number,
          md?.customer_phone ?? null,
          merged.custom_category,
          md?.characteristics ?? null,
          md?.receiver_id ?? null,
          result.transaction_summary.total_customer_paid,
          result.transaction_summary.isolated_tip,
          result.transaction_summary.net_business_remittance,
          result.transaction_summary.shortfall_detected,
          result.employee_financials.calculated_commission,
          result.employee_financials.net_wage_before_tip,
          result.employee_financials.final_payout_output,
          txId,
        ],
      );

      const credit = result.ledger_routing.credit_employee_wages;
      const debit = result.ledger_routing.debit_employee_debt;
      await client.query(
        'UPDATE users SET payable_balance = payable_balance + $1, debt_balance = debt_balance + $2 WHERE id = $3',
        [credit, debit, merged.washer_id],
      );

      const refreshed = (await client.query(`SELECT ${TX_FIELDS} FROM transactions WHERE id = $1`, [txId])).rows[0] as TxRow;
      await syncChaiExpense(client, refreshed, result.transaction_summary.isolated_tip, washer.full_name ?? '');
      await client.query('COMMIT');

      res.json({
        id: txId,
        timestamp: iso(refreshed.timestamp),
        transaction_summary: result.transaction_summary,
        employee_financials: result.employee_financials,
        ledger_routing: result.ledger_routing,
      });
    } catch (e) {
      await client.query('ROLLBACK');
      throw e;
    } finally {
      client.release();
    }
  }),
);

// DELETE /transactions/:id
router.delete(
  '/:id',
  wrap(async (req: AuthRequest, res) => {
    await requireRole(req, ROLES.slice(0, 2));
    const txId = Number(req.params.id);
    if (!Number.isInteger(txId)) throw new HttpError(404, 'Transaction not found');
    const tx = await queryOne<TxRow>(`SELECT ${TX_FIELDS} FROM transactions WHERE id = $1`, [txId]);
    if (!tx) throw new HttpError(404, 'Transaction not found');

    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      await reverseBalances(client, tx);
      await client.query('DELETE FROM expenses WHERE transaction_id = $1', [txId]);
      await client.query('DELETE FROM transactions WHERE id = $1', [txId]);
      await client.query('COMMIT');
    } catch (e) {
      await client.query('ROLLBACK');
      throw e;
    } finally {
      client.release();
    }
    res.json({ message: 'Transaction deleted' });
  }),
);

// Merge existing transaction with a TransactionUpdate into a CreateData.
function mergeUpdate(tx: TxRow, u: Record<string, any>): CreateData {
  const present = <T>(k: string, or: T): T => (u[k] !== undefined ? (u[k] as T) : or);
  const data: CreateData = {
    washer_id: asOptInt(present('washer_id', tx.washer_id), 'washer_id')!,
    category: requireEnum(
      present('category', tx.category.toLowerCase()),
      SERVICE_CATEGORIES,
      'category',
    ),
    expected_price: asOptFloat(present('expected_price', tx.expected_price), 'expected_price')!,
    cash_paid: asOptFloat(present('cash_paid', tx.cash_paid), 'cash_paid')!,
    mpesa_paid: asOptFloat(present('mpesa_paid', tx.mpesa_paid), 'mpesa_paid')!,
    mpesa_transaction_id: present('mpesa_transaction_id', tx.mpesa_transaction_id) ?? null,
    mpesa_sender_name: present('mpesa_sender_name', tx.mpesa_sender_name) ?? null,
    manual_tip: asOptFloat(present('manual_tip', tx.manual_tip), 'manual_tip') ?? 0,
    tip_method: requireEnum(
      present('tip_method', tx.tip_method.toLowerCase()),
      TIP_METHODS,
      'tip_method',
    ),
    misc_amount: asOptFloat(present('misc_amount', tx.misc_amount ?? 0), 'misc_amount') ?? 0,
    misc_description: present('misc_description', tx.misc_description) ?? null,
    has_car_wash: asBool(present('has_car_wash', tx.has_car_wash), 'has_car_wash'),
    has_vacuum: asBool(present('has_vacuum', tx.has_vacuum), 'has_vacuum'),
    has_engine_wash: asBool(present('has_engine_wash', tx.has_engine_wash), 'has_engine_wash'),
    plate_number: present('plate_number', tx.plate_number) ?? null,
    custom_category: present('custom_category', tx.custom_category) ?? null,
    carpet_metadata: tx.customer_phone || tx.carpet_characteristics || tx.receiver_id
      ? {
          characteristics: tx.carpet_characteristics ?? '',
          receiver_id: tx.receiver_id ?? 0,
          customer_phone: tx.customer_phone,
        }
      : null,
  };
  for (const k of ['expected_price', 'cash_paid', 'mpesa_paid', 'manual_tip', 'misc_amount']) {
    if ((data as any)[k] < 0) throw new HttpError(422, `${k} must be >= 0`);
  }
  return data;
}

export default router;